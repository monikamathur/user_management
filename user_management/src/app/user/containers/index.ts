export * from './collection-page.component';
export * from './find-user-page.component';
export * from './selected-user-page.component';
export * from './view-user-page.component';
