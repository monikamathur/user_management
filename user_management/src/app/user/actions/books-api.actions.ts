import { createAction, props } from '@ngrx/store';

import { User } from '@example-app/user/models';

export const searchSuccess = createAction(
  '[Users/API] Search Success',
  props<{ user: User[] }>()
);

export const searchFailure = createAction(
  '[Users/API] Search Failure',
  props<{ errorMsg: string }>()
);
