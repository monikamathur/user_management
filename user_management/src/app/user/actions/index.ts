import * as UserActions from './user.actions';
import * as UsersApiActions from './user-api.actions';
import * as CollectionApiActions from './collection-api.actions';
import * as CollectionPageActions from './collection-page.actions';
import * as FindUserPageActions from './find-user-page.actions';
import * as SelectedUserPageActions from './selected-user-page.actions';
import * as ViewUserPageActions from './view-user-page.actions';

export {
  UserActions,
  UsersApiActions,
  CollectionApiActions,
  CollectionPageActions,
  FindUserPageActions,
  SelectedUserPageActions,
  ViewUserPageActions,
};
