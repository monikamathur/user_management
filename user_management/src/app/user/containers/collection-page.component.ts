import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';

import { Store } from '@ngrx/store';
import { Observable } from 'rxjs';

import { CollectionPageActions } from '../actions';
import { User } from '../models';
import * as fromUsers from '../reducers';

@Component({
  selector: 'bc-collection-page',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <mat-card>
      <mat-card-title>My Collection</mat-card-title>
    </mat-card>

    <user-preview-list [user]="(user$ | async)!"></user-preview-list>
  `,
  /**
   * Container components are permitted to have just enough styles
   * to bring the view together. If the number of styles grow,
   * consider breaking them out into presentational
   * components.
   */
  styles: [
    `
      mat-card-title {
        display: flex;
        justify-content: center;
      }
    `,
  ],
})
export class CollectionPageComponent implements OnInit {
  user$: Observable<User[]>;

  constructor(private store: Store) {
    this.user$ = store.select(fromUsers.selectUserCollection);
  }

  ngOnInit() {
    this.store.dispatch(CollectionPageActions.enter());
  }
}
