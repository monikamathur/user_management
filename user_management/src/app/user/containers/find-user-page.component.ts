import { ChangeDetectionStrategy, Component } from '@angular/core';

import { Store } from '@ngrx/store';
import { Observable } from 'rxjs';
import { take } from 'rxjs/operators';

import { FindUserPageActions } from '../actions';
import { User } from '../models';
import * as fromUsers from '../reducers';

@Component({
  selector: 'bc-find-user-page',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <user-search
      [query]="(searchQuery$ | async)!"
      [searching]="(loading$ | async)!"
      [error]="(error$ | async)!"
      (search)="search($event)"
    >
    </user-search>
    <user-preview-list [user]="(user$ | async)!"> </user-preview-list>
  `,
})
export class FindUserPageComponent {
  searchQuery$: Observable<string>;
  user$: Observable<User[]>;
  loading$: Observable<boolean>;
  error$: Observable<string>;

  constructor(private store: Store) {
    this.searchQuery$ = store.select(fromUsers.selectSearchQuery).pipe(take(1));
    this.user$ = store.select(fromUsers.selectSearchResults);
    this.loading$ = store.select(fromUsers.selectSearchLoading);
    this.error$ = store.select(fromUsers.selectSearchError);
  }

  search(query: string) {
    this.store.dispatch(FindUserPageActions.searchUsers({ query }));
  }
}
