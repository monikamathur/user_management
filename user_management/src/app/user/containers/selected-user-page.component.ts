import { ChangeDetectionStrategy, Component } from '@angular/core';

import { Store } from '@ngrx/store';
import { Observable } from 'rxjs';

import { SelectedUserPageActions } from '../actions';
import { User } from '../models';
import * as fromUsers from '../reducers';

@Component({
  selector: 'selected-user-page',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <user-detail
      [user]="(user$ | async)!"
      [inCollection]="(isSelectedUserInCollection$ | async)!"
      (add)="addToCollection($event)"
      (remove)="removeFromCollection($event)"
    >
    </user-detail>
  `,
})
export class SelectedUserPageComponent {
  user$: Observable<User>;
  isSelectedUserInCollection$: Observable<boolean>;

  constructor(private store: Store) {
    this.user$ = store.select(fromUsers.selectSelectedUser) as Observable<User>;
    this.isSelectedUserInCollection$ = store.select(
      fromUsers.isSelectedUserInCollection
    );
  }

  addToCollection(user: User) {
    this.store.dispatch(SelectedUserPageActions.addUser({ user }));
  }

  removeFromCollection(user: User) {
    this.store.dispatch(SelectedUserPageActions.removeUser({ user }));
  }
}
