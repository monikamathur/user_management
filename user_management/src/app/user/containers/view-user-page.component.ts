import { Component, OnDestroy, ChangeDetectionStrategy } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { Store } from '@ngrx/store';
import { Subscription } from 'rxjs';
import { map } from 'rxjs/operators';

import { ViewUserPageActions } from '../actions';

/**
 * Note: Container components are also reusable. Whether or not
 * a component is a presentation component or a container
 * component is an implementation detail.
 *
 * The View User Page's responsibility is to map router params
 * to a 'Select' user action. Actually showing the selected
 * user remains a responsibility of the
 * SelectedUserPageComponent
 */
@Component({
  selector: 'view-user-page',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: ` <selected-user-page></selected-user-page> `,
})
export class ViewUserPageComponent implements OnDestroy {
  actionsSubscription: Subscription;

  constructor(store: Store, route: ActivatedRoute) {
    this.actionsSubscription = route.params
      .pipe(map((params) => ViewUserPageActions.selectUser({ id: params.id })))
      .subscribe((action) => store.dispatch(action));
  }

  ngOnDestroy() {
    this.actionsSubscription.unsubscribe();
  }
}
