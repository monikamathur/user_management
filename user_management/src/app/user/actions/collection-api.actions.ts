import { createAction, props } from '@ngrx/store';

import { User } from '@example-app/user/models';

/**
 * Add User to Collection Actions
 */
export const addUserSuccess = createAction(
  '[Collection/API] Add User Success',
  props<{ user: User }>()
);

export const addUserFailure = createAction(
  '[Collection/API] Add User Failure',
  props<{ user: User }>()
);

/**
 * Remove User from Collection Actions
 */
export const removeUserSuccess = createAction(
  '[Collection/API] Remove User Success',
  props<{ user: User }>()
);

export const removeUserFailure = createAction(
  '[Collection/API] Remove User Failure',
  props<{ user: User }>()
);

/**
 * Load Collection Actions
 */
export const loadUsersSuccess = createAction(
  '[Collection/API] Load Users Success',
  props<{ user: User[] }>()
);

export const loadUsersFailure = createAction(
  '[Collection/API] Load Users Failure',
  props<{ error: any }>()
);
