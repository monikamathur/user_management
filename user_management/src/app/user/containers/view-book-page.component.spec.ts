import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ActivatedRoute } from '@angular/router';

import { provideMockStore, MockStore } from '@ngrx/store/testing';
import { BehaviorSubject } from 'rxjs';

import {
  UserAuthorsComponent,
  UserDetailComponent,
} from '@example-app/user/components';
import { SelectedUserPageComponent } from '@example-app/user/containers';
import { ViewUserPageComponent } from '@example-app/user/containers';
import { ViewUserPageActions } from '@example-app/user/actions';
import { AddCommasPipe } from '@example-app/shared/pipes/add-commas.pipe';
import { MaterialModule } from '@example-app/material';

describe('View User Page', () => {
  let fixture: ComponentFixture<ViewUserPageComponent>;
  let store: MockStore;
  let route: ActivatedRoute;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [MaterialModule],
      providers: [
        {
          provide: ActivatedRoute,
          useValue: { params: new BehaviorSubject({}) },
        },
        provideMockStore(),
      ],
      declarations: [
        ViewUserPageComponent,
        SelectedUserPageComponent,
        UserDetailComponent,
        UserAuthorsComponent,
        AddCommasPipe,
      ],
    });

    fixture = TestBed.createComponent(ViewUserPageComponent);
    store = TestBed.inject(MockStore);
    route = TestBed.inject(ActivatedRoute);

    jest.spyOn(store, 'dispatch');
  });

  it('should compile', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch a user.Select action on init', () => {
    const action = ViewUserPageActions.selectUser({ id: '2' });

    (route.params as BehaviorSubject<any>).next({ id: '2' });

    expect(store.dispatch).toHaveBeenLastCalledWith(action);
  });
});
