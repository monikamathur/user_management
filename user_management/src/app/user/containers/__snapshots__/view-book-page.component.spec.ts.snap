// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View User Page should compile 1`] = `
<bc-view-user-page
  __ngContext__="0"
  actionsSubscription={[Function SafeSubscriber]}
>
  <bc-selected-user-page>
    <bc-user-detail>
    </bc-user-detail>
  </bc-selected-user-page>
</bc-view-user-page>
`;
